x86/pae.c

```diff
@@ -0,0 +1,267 @@
+		:
+		: "r"(data)
+		: "memory"
+	);
+}
+
+static void paging_enable_pae(void)
+{
+	CRx_TYPE cr0;
+	CRx_TYPE cr4;
+
+	/* Enable PAE */
+	cr4 = read_cr4();
+
+	cr4 |= CR4_PAE;
+	write_cr4(cr4);
+
+	/* Enable Paging */
+	cr0 = read_cr0();
+	cr0 |= CR0_PG;
+	write_cr0(cr0);
+}
+
+static void paging_enable_pae_cr3(uintptr_t cr3)
+{
+	/* Load the page table address */
+	write_cr3(cr3);
+	paging_enable_pae();
+}
+
+static void paging_disable_pae(void)
+{
+	CRx_TYPE cr0;
+	CRx_TYPE cr4;
+
+	/* Disable Paging */
+	cr0 = read_cr0();
+	cr0 &= ~(CRx_TYPE)CR0_PG;
+	write_cr0(cr0);
+
+	/* Disable PAE */
+	cr4 = read_cr4();
+	cr4 &= ~(CRx_TYPE)CR4_PAE;
+	write_cr4(cr4);
+}
+
+/*
+ * Initialize page table instance pgtbl with an identity map of the 32-bit
+ * address space.
+ */
+static void identity_paging_enable(void)
+{
+	struct pde *pd = (struct pde *)&pgtbl.pd, *pdp = (struct pde *)&pgtbl.pdp;
+	/* Point the page directory pointers at the page directories. */
+	memset(&pgtbl.pdp, 0, sizeof(pgtbl.pdp));
+
+	pdp[0].addr_lo = (virt_to_phys((uintptr_t)&pd[512*0])) | PDPTE_PRES;
+	pdp[1].addr_lo = (virt_to_phys((uintptr_t)&pd[512*1])) | PDPTE_PRES;
+	pdp[2].addr_lo = (virt_to_phys((uintptr_t)&pd[512*2])) | PDPTE_PRES;
+	pdp[3].addr_lo = (virt_to_phys((uintptr_t)&pd[512*3])) | PDPTE_PRES;
+
+	for (size_t i = 0; i < 2048; i++) {
+		pd[i].addr_lo = (i << PDE_IDX_SHIFT) | PDE_PS | PDE_PRES | PDE_RW;
+		pd[i].addr_hi = 0;
+	}
+}
+
+/*
+ * Return address outside of filo _start .. _end relocated range, i.e. a linear
+ * address which is not used by FILO CS/DS. Basically any address is fine for
+ * the vmem window use case, as long it does not overlap with relocated FILO
+ * and it is 2 MiB aligned. Use FILO's _start address before relocation.
+ * */
+static void *get_vmem_addr()
+{
+	extern char _start[];
+	const uint32_t addr = (uint32_t)&_start;
+
+	return (void *)ALIGN_UP(addr, s2MiB);
+}
+
+/*
+ * Map a physical page with address phys at virtual address virt, then
+ * invalidate virtual address. The physical address must be 2 MiB aligned.
+ */
+static void map_page(struct pde *const pd, uint64_t phys, void *const virt)
+{
+	pd->addr_lo = phys | PDE_PS | PDE_RW | PDE_PRES;
+	pd->addr_hi = phys >> 32;
+	asm volatile ("invlpg (%0)" :: "b"(virt) : "memory");
+}
+
+void memset_pae(uint64_t dest, unsigned char pat, uint64_t length)
+{
+	ssize_t offset;
+	void *vmem_addr = get_vmem_addr();
+	struct pde *pd;
+
+	if (!pgtbl_initialized) {
+		identity_paging_enable();
+		pgtbl_initialized = 1;
+	}
+
+	paging_enable_pae_cr3(virt_to_phys((uintptr_t)&pgtbl.pdp));
+
+	pd = &pgtbl.pd[(uintptr_t)vmem_addr >> PDE_IDX_SHIFT];
+	offset = dest - ALIGN_DOWN(dest, s2MiB);
+	dest = ALIGN_DOWN(dest, s2MiB);
+
+	do {
+		const size_t len = MIN(length, s2MiB - offset);
+
+		map_page(pd, dest, vmem_addr);
+		debug("%s: Mapped 0x%llx[0x%lx] - 0x%zx\n", __func__,
+				dest + offset, (uintptr_t)vmem_addr + offset, len);
+
+		memset(phys_to_virt(vmem_addr + offset), pat, len);
```

Comment:
If `vmem_addr` is a virtual address as the name implies, why do we need to use `phys_to_virt` here?

Reply:
To neutralize the effect of segmentation. The virtual address is first translated to a linear address by segmentation,when the actual memory is accessed, so we need to add the offset here.

Reply:
I think this should be added as a comment somewhere.

Reply:
I tried to do this in `pae.h` by stating that the given address is a physical address.

The memset call is actually identical to the `memset` call in the non PAE case (see `csl.c`, line 222).

Reply:
But that is the part that tripped me up: the comment of **memset_pae** says physical address and here we use **phys_to_virt** which would suggest we are converting vmem_addr+offset from physical to virtual where we should pass a physical address.